Error when enabling ecAudit in live system

In the scenario where ecAudit is enabled in a live system by restarting one Cassandra node at a time, requests will fail temporarily. The error is manifesting itself as several rejected requests with AuthenticationException at the client side. This goes on for ~10 seconds just after the first Cassandra node is restarted. After this, traffic will work as expected. Subsequent Cassandra node restarts will not cause any issues. The issue was observed in a multi DC cluster.

The following stack trace is collected from failing node:

```
2019-01-25T11:45:38.760+0100 [SharedPool-Worker-14] ERROR o.a.c.t.Message$UnexpectedChannelExceptionHandler:625 apply Unexpected exception during request; channel = [id: 0x2ee41da4, L:/10.61.186.80:12742 - R:/10.61.186.78:33773]
com.google.common.util.concurrent.UncheckedExecutionException: java.lang.RuntimeException: org.apache.cassandra.exceptions.ReadTimeoutException: Operation timed out - received only 3 responses.
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2203) ~[guava-18.0.jar:na]
        at com.google.common.cache.LocalCache.get(LocalCache.java:3937) ~[guava-18.0.jar:na]
        at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3941) ~[guava-18.0.jar:na]
        at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4824) ~[guava-18.0.jar:na]
        at org.apache.cassandra.auth.AuthCache.get(AuthCache.java:108) ~[apache-cassandra-3.0.17-E001.jar:3.0.17-E001]
        at org.apache.cassandra.auth.RolesCache.getRoles(RolesCache.java:44) ~[apache-cassandra-3.0.17-E001.jar:3.0.17-E001]
        at org.apache.cassandra.auth.Roles.getRoles(Roles.java:38) ~[apache-cassandra-3.0.17-E001.jar:3.0.17-E001]
        at com.ericsson.bss.cassandra.ecaudit.filter.role.RoleAuditFilter.isFiltered(RoleAuditFilter.java:46) ~[ecaudit_c3.0-1.0.0.jar:na]
        at com.ericsson.bss.cassandra.ecaudit.filter.yamlandrole.YamlAndRoleAuditFilter.isFiltered(YamlAndRoleAuditFilter.java:48) ~[ecaudit_c3.0-1.0.0.jar:na]
        at com.ericsson.bss.cassandra.ecaudit.facade.DefaultAuditor.audit(DefaultAuditor.java:46) ~[ecaudit_c3.0-1.0.0.jar:na]
        at com.ericsson.bss.cassandra.ecaudit.AuditAdapter.auditAuth(AuditAdapter.java:178) ~[ecaudit_c3.0-1.0.0.jar:na]
        at com.ericsson.bss.cassandra.ecaudit.auth.AuditPasswordAuthenticator$AuditPlainTextSaslAuthenticator.getAuthenticatedUser(AuditPasswordAuthenticator.java:140) ~[ecaudit_c3.0-1.0.0.jar:na]
        at org.apache.cassandra.transport.messages.AuthResponse.execute(AuthResponse.java:78) ~[apache-cassandra-3.0.17-E001.jar:3.0.17-E001]
        at org.apache.cassandra.transport.Message$Dispatcher.channelRead0(Message.java:513) [apache-cassandra-3.0.17-E001.jar:3.0.17-E001]
        at org.apache.cassandra.transport.Message$Dispatcher.channelRead0(Message.java:407) [apache-cassandra-3.0.17-E001.jar:3.0.17-E001]
        at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105) [netty-all-4.0.44.Final.jar:4.0.44.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:357) [netty-all-4.0.44.Final.jar:4.0.44.Final]
        at io.netty.channel.AbstractChannelHandlerContext.access$600(AbstractChannelHandlerContext.java:35) [netty-all-4.0.44.Final.jar:4.0.44.Final]
        at io.netty.channel.AbstractChannelHandlerContext$7.run(AbstractChannelHandlerContext.java:348) [netty-all-4.0.44.Final.jar:4.0.44.Final]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_192]
        at org.apache.cassandra.concurrent.AbstractLocalAwareExecutorService$FutureTask.run(AbstractLocalAwareExecutorService.java:164) [apache-cassandra-3.0.17-E001.jar:3.0.17-E001]
        at org.apache.cassandra.concurrent.SEPWorker.run(SEPWorker.java:105) [apache-cassandra-3.0.17-E001.jar:3.0.17-E001]
        at java.lang.Thread.run(Thread.java:748) [na:1.8.0_192]
Caused by: java.lang.RuntimeException: org.apache.cassandra.exceptions.ReadTimeoutException: Operation timed out - received only 3 responses.
        at org.apache.cassandra.auth.CassandraRoleManager.getRole(CassandraRoleManager.java:517) ~[apache-cassandra-3.0.17-E001.jar:3.0.17-E001]
        at org.apache.cassandra.auth.CassandraRoleManager.getRoles(CassandraRoleManager.java:282) ~[apache-cassandra-3.0.17-E001.jar:3.0.17-E001]
        at org.apache.cassandra.auth.RolesCache.lambda$new$0(RolesCache.java:36) ~[apache-cassandra-3.0.17-E001.jar:3.0.17-E001]
        at org.apache.cassandra.auth.AuthCache$1.load(AuthCache.java:183) ~[apache-cassandra-3.0.17-E001.jar:3.0.17-E001]
        at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3527) ~[guava-18.0.jar:na]
        at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2319) ~[guava-18.0.jar:na]
        at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2282) ~[guava-18.0.jar:na]
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2197) ~[guava-18.0.jar:na]
        ... 22 common frames omitted
Caused by: org.apache.cassandra.exceptions.ReadTimeoutException: Operation timed out - received only 3 responses.
        at org.apache.cassandra.service.ReadCallback.awaitResults(ReadCallback.java:132) ~[apache-cassandra-3.0.17-E001.jar:3.0.17-E001]
        at org.apache.cassandra.service.ReadCallback.get(ReadCallback.java:137) ~[apache-cassandra-3.0.17-E001.jar:3.0.17-E001]
        at org.apache.cassandra.service.AbstractReadExecutor.get(AbstractReadExecutor.java:145) ~[apache-cassandra-3.0.17-E001.jar:3.0.17-E001]
        at org.apache.cassandra.service.StorageProxy$SinglePartitionReadLifecycle.awaitResultsAndRetryOnDigestMismatch(StorageProxy.java:1716) ~[apache-cassandra-3.0.17-E001.jar:3.0.17-E001]
        at org.apache.cassandra.service.StorageProxy.fetchRows(StorageProxy.java:1665) ~[apache-cassandra-3.0.17-E001.jar:3.0.17-E001]
        at org.apache.cassandra.service.StorageProxy.readRegular(StorageProxy.java:1603) ~[apache-cassandra-3.0.17-E001.jar:3.0.17-E001]
        at org.apache.cassandra.service.StorageProxy.read(StorageProxy.java:1522) ~[apache-cassandra-3.0.17-E001.jar:3.0.17-E001]
        at org.apache.cassandra.db.SinglePartitionReadCommand$Group.execute(SinglePartitionReadCommand.java:1133) ~[apache-cassandra-3.0.17-E001.jar:3.0.17-E001]
        at org.apache.cassandra.cql3.statements.SelectStatement.execute(SelectStatement.java:303) ~[apache-cassandra-3.0.17-E001.jar:3.0.17-E001]
        at org.apache.cassandra.cql3.statements.SelectStatement.execute(SelectStatement.java:264) ~[apache-cassandra-3.0.17-E001.jar:3.0.17-E001]
        at org.apache.cassandra.auth.CassandraRoleManager.getRoleFromTable(CassandraRoleManager.java:525) ~[apache-cassandra-3.0.17-E001.jar:3.0.17-E001]
        at org.apache.cassandra.auth.CassandraRoleManager.getRole(CassandraRoleManager.java:507) ~[apache-cassandra-3.0.17-E001.jar:3.0.17-E001]
        ... 29 common frames omitted
```

Can reproduce this by:
- create ccm cluster with 2+2 nodes
- configure NWTopology on system_auth with RF=2+2
- repair system_auth
- execute cassandra_stress with slow steady state
- alter keyspace1 RF=2+2
- ccm node1 stop
- ecaudit/bin/install_into_current_ccm.sh
- ccm node1 start

Will get two different kinds of errors in the log of node1.
This:
```
com.google.common.util.concurrent.UncheckedExecutionException: org.apache.cassandra.exceptions.ReadTimeoutException: Operation timed out - received only 1 responses.
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2203) ~[guava-18.0.jar:na]
        at com.google.common.cache.LocalCache.get(LocalCache.java:3937) ~[guava-18.0.jar:na]
        at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3941) ~[guava-18.0.jar:na]
        at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4824) ~[guava-18.0.jar:na]
        at com.ericsson.bss.cassandra.ecaudit.auth.cache.AuthCache.get(AuthCache.java:112) ~[ecaudit_c3.0-2.1.0-SNAPSHOT.jar:na]
        at com.ericsson.bss.cassandra.ecaudit.auth.AuditWhitelistCache.getWhitelist(AuditWhitelistCache.java:70) ~[ecaudit_c3.0-2.1.0-SNAPSHOT.jar:na]
        at com.ericsson.bss.cassandra.ecaudit.filter.role.RoleAuditFilter.isOperationWhitelistedOnResourceByRole(RoleAuditFilter.java:125) ~[ecaudit_c3.0-2.1.0-SNAPSHOT.jar:na]
        at com.ericsson.bss.cassandra.ecaudit.filter.role.RoleAuditFilter.isOperationWhitelistedOnResourceByRoles(RoleAuditFilter.java:114) ~[ecaudit_c3.0-2.1.0-SNAPSHOT.jar:na]
        at com.ericsson.bss.cassandra.ecaudit.filter.role.RoleAuditFilter.isFiltered(RoleAuditFilter.java:88) ~[ecaudit_c3.0-2.1.0-SNAPSHOT.jar:na]
        at com.ericsson.bss.cassandra.ecaudit.filter.role.RoleAuditFilter.isFiltered(RoleAuditFilter.java:79) ~[ecaudit_c3.0-2.1.0-SNAPSHOT.jar:na]
        at com.ericsson.bss.cassandra.ecaudit.facade.DefaultAuditor.audit(DefaultAuditor.java:51) ~[ecaudit_c3.0-2.1.0-SNAPSHOT.jar:na]
        at com.ericsson.bss.cassandra.ecaudit.AuditAdapter.auditAuth(AuditAdapter.java:188) ~[ecaudit_c3.0-2.1.0-SNAPSHOT.jar:na]
        at com.ericsson.bss.cassandra.ecaudit.auth.AuditPasswordAuthenticator$AuditPlainTextSaslAuthenticator.getAuthenticatedUser(AuditPasswordAuthenticator.java:132) ~[ecaudit_c3.0-2.1.0-SNAPSHOT.jar:na]
        at org.apache.cassandra.transport.messages.AuthResponse.execute(AuthResponse.java:78) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.transport.Message$Dispatcher.channelRead0(Message.java:562) [apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.transport.Message$Dispatcher.channelRead0(Message.java:407) [apache-cassandra-3.0.18.jar:3.0.18]
        at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105) [netty-all-4.0.44.Final.jar:4.0.44.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:357) [netty-all-4.0.44.Final.jar:4.0.44.Final]
        at io.netty.channel.AbstractChannelHandlerContext.access$600(AbstractChannelHandlerContext.java:35) [netty-all-4.0.44.Final.jar:4.0.44.Final]
        at io.netty.channel.AbstractChannelHandlerContext$7.run(AbstractChannelHandlerContext.java:348) [netty-all-4.0.44.Final.jar:4.0.44.Final]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_181]
        at org.apache.cassandra.concurrent.AbstractLocalAwareExecutorService$FutureTask.run(AbstractLocalAwareExecutorService.java:164) [apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.concurrent.SEPWorker.run(SEPWorker.java:110) [apache-cassandra-3.0.18.jar:3.0.18]
        at java.lang.Thread.run(Thread.java:748) [na:1.8.0_181]
Caused by: org.apache.cassandra.exceptions.ReadTimeoutException: Operation timed out - received only 1 responses.
        at org.apache.cassandra.service.ReadCallback.awaitResults(ReadCallback.java:132) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.service.ReadCallback.get(ReadCallback.java:137) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.service.AbstractReadExecutor.get(AbstractReadExecutor.java:145) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.service.StorageProxy$SinglePartitionReadLifecycle.awaitResultsAndRetryOnDigestMismatch(StorageProxy.java:1704) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.service.StorageProxy.fetchRows(StorageProxy.java:1653) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.service.StorageProxy.readRegular(StorageProxy.java:1591) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.service.StorageProxy.read(StorageProxy.java:1510) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.db.SinglePartitionReadCommand$Group.execute(SinglePartitionReadCommand.java:1179) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.cql3.statements.SelectStatement.execute(SelectStatement.java:303) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.cql3.statements.SelectStatement.execute(SelectStatement.java:264) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at com.ericsson.bss.cassandra.ecaudit.auth.WhitelistDataAccess.getWhitelist(WhitelistDataAccess.java:140) ~[ecaudit_c3.0-2.1.0-SNAPSHOT.jar:na]
        at com.ericsson.bss.cassandra.ecaudit.auth.cache.AuthCache$1.load(AuthCache.java:187) ~[ecaudit_c3.0-2.1.0-SNAPSHOT.jar:na]
        at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3527) ~[guava-18.0.jar:na]
        at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2319) ~[guava-18.0.jar:na]
        at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2282) ~[guava-18.0.jar:na]
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2197) ~[guava-18.0.jar:na]
        ... 23 common frames omitted

```
And this:
```
ERROR [SharedPool-Worker-3] 2019-02-26 14:55:59,302 Message.java:689 - Unexpected exception during request; channel = [id: 0x0e81fc9b, L:/127.0.0.1:9042 - R:/127.0.0.1:39294]
com.google.common.util.concurrent.UncheckedExecutionException: java.lang.RuntimeException: org.apache.cassandra.exceptions.ReadTimeoutException: Operation timed out - received only 2 responses.
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2203) ~[guava-18.0.jar:na]
        at com.google.common.cache.LocalCache.get(LocalCache.java:3937) ~[guava-18.0.jar:na]
        at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3941) ~[guava-18.0.jar:na]
        at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4824) ~[guava-18.0.jar:na]
        at org.apache.cassandra.auth.RolesCache.getRoles(RolesCache.java:63) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.auth.Roles.getRoles(Roles.java:38) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at com.ericsson.bss.cassandra.ecaudit.filter.role.RoleAuditFilter.isFiltered(RoleAuditFilter.java:79) ~[ecaudit_c3.0-2.1.0-SNAPSHOT.jar:na]
        at com.ericsson.bss.cassandra.ecaudit.facade.DefaultAuditor.audit(DefaultAuditor.java:51) ~[ecaudit_c3.0-2.1.0-SNAPSHOT.jar:na]
        at com.ericsson.bss.cassandra.ecaudit.AuditAdapter.auditAuth(AuditAdapter.java:188) ~[ecaudit_c3.0-2.1.0-SNAPSHOT.jar:na]
        at com.ericsson.bss.cassandra.ecaudit.auth.AuditPasswordAuthenticator$AuditPlainTextSaslAuthenticator.getAuthenticatedUser(AuditPasswordAuthenticator.java:132) ~[ecaudit_c3.0-2.1.0-SNAPSHOT.jar:na]
        at org.apache.cassandra.transport.messages.AuthResponse.execute(AuthResponse.java:78) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.transport.Message$Dispatcher.channelRead0(Message.java:562) [apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.transport.Message$Dispatcher.channelRead0(Message.java:407) [apache-cassandra-3.0.18.jar:3.0.18]
        at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105) [netty-all-4.0.44.Final.jar:4.0.44.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:357) [netty-all-4.0.44.Final.jar:4.0.44.Final]
        at io.netty.channel.AbstractChannelHandlerContext.access$600(AbstractChannelHandlerContext.java:35) [netty-all-4.0.44.Final.jar:4.0.44.Final]
        at io.netty.channel.AbstractChannelHandlerContext$7.run(AbstractChannelHandlerContext.java:348) [netty-all-4.0.44.Final.jar:4.0.44.Final]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_181]
        at org.apache.cassandra.concurrent.AbstractLocalAwareExecutorService$FutureTask.run(AbstractLocalAwareExecutorService.java:164) [apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.concurrent.SEPWorker.run(SEPWorker.java:110) [apache-cassandra-3.0.18.jar:3.0.18]
        at java.lang.Thread.run(Thread.java:748) [na:1.8.0_181]
Caused by: java.lang.RuntimeException: org.apache.cassandra.exceptions.ReadTimeoutException: Operation timed out - received only 2 responses.
        at org.apache.cassandra.auth.CassandraRoleManager.getRole(CassandraRoleManager.java:517) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.auth.CassandraRoleManager.getRoles(CassandraRoleManager.java:282) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at com.ericsson.bss.cassandra.ecaudit.auth.AuditRoleManager.getRoles(AuditRoleManager.java:173) ~[ecaudit_c3.0-2.1.0-SNAPSHOT.jar:na]
        at org.apache.cassandra.auth.RolesCache$1.load(RolesCache.java:115) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.auth.RolesCache$1.load(RolesCache.java:112) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3527) ~[guava-18.0.jar:na]
        at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2319) ~[guava-18.0.jar:na]
        at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2282) ~[guava-18.0.jar:na]
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2197) ~[guava-18.0.jar:na]
        ... 20 common frames omitted
Caused by: org.apache.cassandra.exceptions.ReadTimeoutException: Operation timed out - received only 2 responses.
        at org.apache.cassandra.service.ReadCallback.awaitResults(ReadCallback.java:132) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.service.ReadCallback.get(ReadCallback.java:137) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.service.AbstractReadExecutor.get(AbstractReadExecutor.java:145) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.service.StorageProxy$SinglePartitionReadLifecycle.awaitResultsAndRetryOnDigestMismatch(StorageProxy.java:1704) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.service.StorageProxy.fetchRows(StorageProxy.java:1653) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.service.StorageProxy.readRegular(StorageProxy.java:1591) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.service.StorageProxy.read(StorageProxy.java:1510) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.db.SinglePartitionReadCommand$Group.execute(SinglePartitionReadCommand.java:1179) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.cql3.statements.SelectStatement.execute(SelectStatement.java:303) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.cql3.statements.SelectStatement.execute(SelectStatement.java:264) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.auth.CassandraRoleManager.getRoleFromTable(CassandraRoleManager.java:525) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.auth.CassandraRoleManager.getRole(CassandraRoleManager.java:507) ~[apache-cassandra-3.0.18.jar:3.0.18]
        ... 28 common frames omitted
```
These stack traces keeps coming for a minute or so.

The other nodes have error messages like this:
```
WARN  [MessagingService-Incoming-/127.0.0.1] 2019-02-26 14:55:49,284 IncomingTcpConnection.java:101 - UnknownColumnFamilyException reading from socket; closing
org.apache.cassandra.db.UnknownColumnFamilyException: Couldn't find table for cfId 0aafeebd-0739-3411-9051-31a04105a825. If a table was just created, this is likely due to the schema not being fully propagated.  Please wait for schema agreement on table creation.
        at org.apache.cassandra.config.CFMetaData$Serializer.deserialize(CFMetaData.java:1495) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.db.ReadCommand$Serializer.deserialize(ReadCommand.java:660) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.db.ReadCommand$Serializer.deserialize(ReadCommand.java:599) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.io.ForwardingVersionedSerializer.deserialize(ForwardingVersionedSerializer.java:50) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.net.MessageIn.read(MessageIn.java:98) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.net.IncomingTcpConnection.receiveMessage(IncomingTcpConnection.java:201) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.net.IncomingTcpConnection.receiveMessages(IncomingTcpConnection.java:178) ~[apache-cassandra-3.0.18.jar:3.0.18]
        at org.apache.cassandra.net.IncomingTcpConnection.run(IncomingTcpConnection.java:92) ~[apache-cassandra-3.0.18.jar:3.0.18]
```

On client-side in cassandra-stress, I get this:
```
ERROR 14:56:05,443 Authentication error on host localhost/127.0.0.1:9042: com.google.common.util.concurrent.UncheckedExecutionException: org.apache.cassandra.exceptions.ReadTimeoutException: Operation timed out - received only 1 responses.
ERROR 14:56:05,452 Authentication error during reconnection to localhost/127.0.0.1:9042, scheduling retry in 2000 milliseconds
com.datastax.driver.core.exceptions.AuthenticationException: Authentication error on host localhost/127.0.0.1:9042: com.google.common.util.concurrent.UncheckedExecutionException: org.apache.cassandra.exceptions.ReadTimeoutException: Operation timed out - received only 1 responses.
	at com.datastax.driver.core.Connection$8.apply(Connection.java:377) ~[cassandra-driver-core-3.0.1-shaded.jar:na]
	at com.datastax.driver.core.Connection$8.apply(Connection.java:347) ~[cassandra-driver-core-3.0.1-shaded.jar:na]
	at com.google.common.util.concurrent.Futures$ChainingListenableFuture.run(Futures.java:906) ~[guava-18.0.jar:na]
	at com.google.common.util.concurrent.Futures$1$1.run(Futures.java:635) ~[guava-18.0.jar:na]
	at com.google.common.util.concurrent.MoreExecutors$DirectExecutorService.execute(MoreExecutors.java:299) ~[guava-18.0.jar:na]
	at com.google.common.util.concurrent.Futures$1.run(Futures.java:632) ~[guava-18.0.jar:na]
	at com.google.common.util.concurrent.MoreExecutors$DirectExecutor.execute(MoreExecutors.java:457) ~[guava-18.0.jar:na]
	at com.google.common.util.concurrent.ExecutionList.executeListener(ExecutionList.java:156) ~[guava-18.0.jar:na]
	at com.google.common.util.concurrent.ExecutionList.execute(ExecutionList.java:145) ~[guava-18.0.jar:na]
	at com.google.common.util.concurrent.AbstractFuture.set(AbstractFuture.java:185) ~[guava-18.0.jar:na]
	at com.datastax.driver.core.Connection$Future.onSet(Connection.java:1182) ~[cassandra-driver-core-3.0.1-shaded.jar:na]
	at com.datastax.driver.core.Connection$Dispatcher.channelRead0(Connection.java:1013) ~[cassandra-driver-core-3.0.1-shaded.jar:na]
	at com.datastax.driver.core.Connection$Dispatcher.channelRead0(Connection.java:936) ~[cassandra-driver-core-3.0.1-shaded.jar:na]
	at com.datastax.shaded.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105) ~[cassandra-driver-core-3.0.1-shaded.jar:na]
	at com.datastax.shaded.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:318) ~[cassandra-driver-core-3.0.1-shaded.jar:na]
	at com.datastax.shaded.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:304) ~[cassandra-driver-core-3.0.1-shaded.jar:na]
	at com.datastax.shaded.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:266) ~[cassandra-driver-core-3.0.1-shaded.jar:na]
	at com.datastax.shaded.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:318) ~[cassandra-driver-core-3.0.1-shaded.jar:na]
	at com.datastax.shaded.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:304) ~[cassandra-driver-core-3.0.1-shaded.jar:na]
	at com.datastax.shaded.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103) ~[cassandra-driver-core-3.0.1-shaded.jar:na]
	at com.datastax.shaded.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:318) ~[cassandra-driver-core-3.0.1-shaded.jar:na]
	at com.datastax.shaded.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:304) ~[cassandra-driver-core-3.0.1-shaded.jar:na]
	at com.datastax.shaded.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:276) ~[cassandra-driver-core-3.0.1-shaded.jar:na]
	at com.datastax.shaded.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:263) ~[cassandra-driver-core-3.0.1-shaded.jar:na]
	at com.datastax.shaded.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:318) ~[cassandra-driver-core-3.0.1-shaded.jar:na]
	at com.datastax.shaded.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:304) ~[cassandra-driver-core-3.0.1-shaded.jar:na]
	at com.datastax.shaded.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:846) ~[cassandra-driver-core-3.0.1-shaded.jar:na]
	at com.datastax.shaded.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:131) ~[cassandra-driver-core-3.0.1-shaded.jar:na]
	at com.datastax.shaded.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:511) ~[cassandra-driver-core-3.0.1-shaded.jar:na]
	at com.datastax.shaded.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:468) ~[cassandra-driver-core-3.0.1-shaded.jar:na]
	at com.datastax.shaded.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:382) ~[cassandra-driver-core-3.0.1-shaded.jar:na]
	at com.datastax.shaded.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:354) ~[cassandra-driver-core-3.0.1-shaded.jar:na]
	at com.datastax.shaded.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:112) ~[cassandra-driver-core-3.0.1-shaded.jar:na]
	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_181]
```

--- a/ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/AuditAdapter.java
+++ b/ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/AuditAdapter.java
@@ -33,6 +33,8 @@ import org.apache.cassandra.cql3.BatchQueryOptions;
 import org.apache.cassandra.cql3.CQLStatement;
 import org.apache.cassandra.cql3.QueryOptions;
 import org.apache.cassandra.cql3.statements.BatchStatement;
+import org.apache.cassandra.exceptions.AuthenticationException;
+import org.apache.cassandra.exceptions.RequestExecutionException;
 import org.apache.cassandra.service.ClientState;
 import org.apache.cassandra.utils.MD5Digest;
 
@@ -160,12 +162,13 @@ public class AuditAdapter
     /**
      * Audit an authentication attempt.
      *
-     * @param username the user to authenticate
-     * @param clientIp the address of the client that tries to authenticate
-     * @param status   the status of the operation
+     * @param username  the user to authenticate
+     * @param clientIp  the address of the client that tries to authenticate
+     * @param status    the status of the operation
      * @param timestamp the system timestamp for the request
+     * @throws AuthenticationException if the audit operation could not be performed
      */
-    public void auditAuth(String username, InetAddress clientIp, Status status, long timestamp)
+    public void auditAuth(String username, InetAddress clientIp, Status status, long timestamp) throws AuthenticationException
     {
         AuditEntry logEntry = entryBuilderFactory.createAuthenticationEntryBuilder()
                                                  .client(clientIp)
@@ -175,7 +178,14 @@ public class AuditAdapter
                                                  .timestamp(timestamp)
                                                  .build();
 
-        auditor.audit(logEntry);
+        try
+        {
+            auditor.audit(logEntry);
+        }
+        catch (RequestExecutionException e)
+        {
+            throw new AuthenticationException(e.toString());
+        }
     }
 
     /**

--- a/ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/auth/Exceptions.java
+++ b/ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/auth/Exceptions.java
@@ -1,0 +1,60 @@
+/*
+ * Copyright 2019 Telefonaktiebolaget LM Ericsson
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.ericsson.bss.cassandra.ecaudit.auth;
+
+import org.apache.cassandra.exceptions.CassandraException;
+
+public class Exceptions
+{
+    /**
+     * Try to find a CassandraException which is wrapped in a generic RuntimeException.
+     *
+     * This is useful for instance when cache updates fail asynchronously
+     * with a RuntimeException or an UncheckedExecutionException
+     * that is wrapping a CassandraException which we really want to propagate back to the client.
+     *
+     * The unwrapping procedure will cause valuable back trace to be lost,
+     * and so the original exception is added as a suppressed exception to the CassandraException.
+     *
+     * @param exception the generic RuntimeException
+     * @return the CassandraException if found in the wrapped stack, otherwise the original RuntimeException
+     */
+    public static RuntimeException tryGetCassandraExceptionCause(RuntimeException exception)
+    {
+        Throwable cause = exception.getCause();
+        if (cause instanceof CassandraException)
+        {
+            cause.addSuppressed(exception);
+            return (CassandraException) cause;
+        }
+        else if (cause instanceof RuntimeException)
+        {
+            RuntimeException causeOfCause = tryGetCassandraExceptionCause((RuntimeException) cause);
+
+            if (causeOfCause == cause)
+            {
+                return exception;
+            }
+
+            causeOfCause.addSuppressed(exception);
+            return causeOfCause;
+        }
+        else
+        {
+            return exception;
+        }
+    }
+}

--- a/ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/auth/SchemaHelper.java
+++ b/ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/auth/SchemaHelper.java
@@ -1,0 +1,98 @@
+/*
+ * Copyright 2019 Telefonaktiebolaget LM Ericsson
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.ericsson.bss.cassandra.ecaudit.auth;
+
+import java.net.InetAddress;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.gms.ApplicationState;
+import org.apache.cassandra.gms.EndpointState;
+import org.apache.cassandra.gms.Gossiper;
+import org.apache.cassandra.gms.VersionedValue;
+import org.apache.cassandra.utils.FBUtilities;
+
+class SchemaHelper
+{
+    private static final Logger LOG = LoggerFactory.getLogger(SchemaHelper.class);
+
+    private final InetAddress localAddress;
+    private final Gossiper gossiper;
+
+    SchemaHelper()
+    {
+        this(FBUtilities.getBroadcastAddress(), Gossiper.instance);
+    }
+
+    @VisibleForTesting
+    SchemaHelper(InetAddress localAddress, Gossiper gossiper)
+    {
+        this.localAddress = localAddress;
+        this.gossiper = gossiper;
+    }
+
+    void waitForSchemaAlignment(int maxTimeToWaitInSeconds)
+    {
+        LOG.info("Waiting for schema to align in cluster");
+        int timeToWaitInSeconds = maxTimeToWaitInSeconds;
+
+        while(areLiveNodesWaitingOnSchemaUpdate() && timeToWaitInSeconds-- > 0)
+        {
+            try
+            {
+                Thread.sleep(1000);
+            }
+            catch (InterruptedException e)
+            {
+                Thread.currentThread().interrupt();
+                return;
+            }
+        }
+    }
+
+    private boolean areLiveNodesWaitingOnSchemaUpdate()
+    {
+        EndpointState localEndpointState = gossiper.getEndpointStateForEndpoint(localAddress);
+        String localSchemaId = localEndpointState.getApplicationState(ApplicationState.SCHEMA).value;
+
+        for (InetAddress memberAddress : gossiper.getLiveMembers())
+        {
+            if (isRemoteEndpoint(memberAddress))
+            {
+                EndpointState memberEndpointState = gossiper.getEndpointStateForEndpoint(memberAddress);
+                VersionedValue memberVersionedSchema = memberEndpointState.getApplicationState(ApplicationState.SCHEMA);
+                if (isDifferentSchema(localSchemaId, memberVersionedSchema)) {
+                    LOG.debug("Endpoint {} is not aligned yet", memberAddress);
+                    return true;
+                }
+            }
+        }
+
+        return false;
+    }
+
+    private boolean isRemoteEndpoint(InetAddress memberAddress)
+    {
+        return !memberAddress.equals(localAddress);
+    }
+
+    private boolean isDifferentSchema(String localSchema, VersionedValue versionedValue)
+    {
+        return versionedValue != null && !localSchema.equals(versionedValue.value);
+    }
+}

--- a/ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/auth/SchemaHelper.java
+++ b/ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/auth/SchemaHelper.java
@@ -24,60 +24,73 @@ import org.slf4j.LoggerFactory;
 import org.apache.cassandra.gms.ApplicationState;
 import org.apache.cassandra.gms.EndpointState;
 import org.apache.cassandra.gms.Gossiper;
-import org.apache.cassandra.gms.VersionedValue;
 import org.apache.cassandra.utils.FBUtilities;
 
 class SchemaHelper
 {
     private static final Logger LOG = LoggerFactory.getLogger(SchemaHelper.class);
 
+    private static final long RETRY_INTERVAL_MS = 1000;
+
     private final InetAddress localAddress;
     private final Gossiper gossiper;
+    private final long retryIntervalMillis;
 
     SchemaHelper()
     {
-        this(FBUtilities.getBroadcastAddress(), Gossiper.instance);
+        this(FBUtilities.getBroadcastAddress(), Gossiper.instance, RETRY_INTERVAL_MS);
     }
 
     @VisibleForTesting
-    SchemaHelper(InetAddress localAddress, Gossiper gossiper)
+    SchemaHelper(InetAddress localAddress, Gossiper gossiper, long retryIntervalMillis)
     {
         this.localAddress = localAddress;
         this.gossiper = gossiper;
+        this.retryIntervalMillis = retryIntervalMillis;
     }
 
-    void waitForSchemaAlignment(int maxTimeToWaitInSeconds)
+    boolean areSchemasAligned(long schemaAlignmentDelayMillis)
     {
         LOG.info("Waiting for schema to align in cluster");
-        int timeToWaitInSeconds = maxTimeToWaitInSeconds;
 
-        while(areLiveNodesWaitingOnSchemaUpdate() && timeToWaitInSeconds-- > 0)
+        int delayMillis = 0;
+        while (areLiveNodesWaitingOnSchemaUpdate())
         {
+            if (delayMillis >= schemaAlignmentDelayMillis)
+            {
+                return false;
+            }
+            delayMillis += retryIntervalMillis;
+
             try
             {
-                Thread.sleep(1000);
+                Thread.sleep(retryIntervalMillis);
             }
             catch (InterruptedException e)
             {
                 Thread.currentThread().interrupt();
-                return;
+                return false;
             }
         }
+
+        return true;
     }
 
     private boolean areLiveNodesWaitingOnSchemaUpdate()
     {
-        EndpointState localEndpointState = gossiper.getEndpointStateForEndpoint(localAddress);
-        String localSchemaId = localEndpointState.getApplicationState(ApplicationState.SCHEMA).value;
+        String localSchemaId = tryGetSchemaId(localAddress);
+        if (localSchemaId == null)
+        {
+            return true;
+        }
 
         for (InetAddress memberAddress : gossiper.getLiveMembers())
         {
             if (isRemoteEndpoint(memberAddress))
             {
-                EndpointState memberEndpointState = gossiper.getEndpointStateForEndpoint(memberAddress);
-                VersionedValue memberVersionedSchema = memberEndpointState.getApplicationState(ApplicationState.SCHEMA);
-                if (isDifferentSchema(localSchemaId, memberVersionedSchema)) {
-                    LOG.debug("Endpoint {} is not aligned yet", memberAddress);
+                if (isDifferentSchema(localSchemaId, memberAddress))
+                {
+                    LOG.debug("Waiting for schema alignment at endpoint {}", memberAddress);
                     return true;
                 }
             }
@@ -86,13 +99,27 @@ class SchemaHelper
         return false;
     }
 
+    private String tryGetSchemaId(InetAddress address)
+    {
+        EndpointState endpointState = gossiper.getEndpointStateForEndpoint(address);
+        if (endpointState != null)
+        {
+            return endpointState.getApplicationState(ApplicationState.SCHEMA).value;
+        }
+        else
+        {
+            return null;
+        }
+    }
+
     private boolean isRemoteEndpoint(InetAddress memberAddress)
     {
         return !memberAddress.equals(localAddress);
     }
 
-    private boolean isDifferentSchema(String localSchema, VersionedValue versionedValue)
+    private boolean isDifferentSchema(String localSchemaId, InetAddress memberAddress)
     {
-        return versionedValue != null && !localSchema.equals(versionedValue.value);
+        String remoteSchemaId = tryGetSchemaId(memberAddress);
+        return !localSchemaId.equals(remoteSchemaId);
     }
 }

--- a/ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/auth/WhitelistDataAccess.java
+++ b/ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/auth/WhitelistDataAccess.java
@@ -58,6 +58,7 @@ public class WhitelistDataAccess
 {
     private static final Logger LOG = LoggerFactory.getLogger(WhitelistDataAccess.class);
 
+    private static final int MAX_WAIT_TIME_FOR_SCHEMA_ALIGNMENT_IN_SECONDS = 120;
     private boolean setupCompleted = false;
 
     private static final String DEFAULT_SUPERUSER_NAME = "cassandra";
@@ -203,13 +204,21 @@ public class WhitelistDataAccess
         KeyspaceMetadata expected = AuditAuthKeyspace.metadata();
         KeyspaceMetadata defined = Schema.instance.getKSMetaData(expected.name);
 
+        boolean changesAnnounced = false;
         for (CFMetaData expectedTable : expected.tables)
         {
             CFMetaData definedTable = defined.tables.get(expectedTable.cfName).orElse(null);
             if (definedTable == null || !definedTable.equals(expectedTable))
             {
                 MigrationManager.forceAnnounceNewColumnFamily(expectedTable);
+                changesAnnounced = true;
             }
+        }
+
+        if (changesAnnounced)
+        {
+            SchemaHelper schemaHelper = new SchemaHelper();
+            schemaHelper.waitForSchemaAlignment(MAX_WAIT_TIME_FOR_SCHEMA_ALIGNMENT_IN_SECONDS);
         }
     }
 

--- a/ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/auth/WhitelistDataAccess.java
+++ b/ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/auth/WhitelistDataAccess.java
@@ -58,7 +58,8 @@ public class WhitelistDataAccess
 {
     private static final Logger LOG = LoggerFactory.getLogger(WhitelistDataAccess.class);
 
-    private static final int MAX_WAIT_TIME_FOR_SCHEMA_ALIGNMENT_IN_SECONDS = 120;
+    private static final long SCHEMA_ALIGNMENT_DELAY_MS = Long.getLong("ecaudit.schema_alignment_delay_ms", 120_000L);
+
     private boolean setupCompleted = false;
 
     private static final String DEFAULT_SUPERUSER_NAME = "cassandra";
@@ -135,7 +136,7 @@ public class WhitelistDataAccess
         QueryProcessor.process(statement, consistencyForRole(role));
     }
 
-    public Map<IResource, Set<Permission>> getWhitelist(RoleResource role)
+    Map<IResource, Set<Permission>> getWhitelist(RoleResource role)
     {
         ResultMessage.Rows rows = loadWhitelistStatement.execute(
                 QueryState.forInternalCalls(),
@@ -218,7 +219,10 @@ public class WhitelistDataAccess
         if (changesAnnounced)
         {
             SchemaHelper schemaHelper = new SchemaHelper();
-            schemaHelper.waitForSchemaAlignment(MAX_WAIT_TIME_FOR_SCHEMA_ALIGNMENT_IN_SECONDS);
+            if (!schemaHelper.areSchemasAligned(SCHEMA_ALIGNMENT_DELAY_MS))
+            {
+                LOG.warn("Schema alignment timeout - continuing startup");
+            }
         }
     }
 

--- a/ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/filter/role/RoleAuditFilter.java
+++ b/ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/filter/role/RoleAuditFilter.java
@@ -19,6 +19,7 @@ import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.function.Function;
 
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.util.concurrent.UncheckedExecutionException;
@@ -32,7 +33,6 @@ import org.apache.cassandra.auth.IResource;
 import org.apache.cassandra.auth.Permission;
 import org.apache.cassandra.auth.RoleResource;
 import org.apache.cassandra.auth.Roles;
-import org.apache.cassandra.exceptions.CassandraException;
 
 /**
  * A role based white-list filter that exempts users based on custom options on roles in Cassandra.
@@ -41,17 +41,19 @@ public class RoleAuditFilter implements AuditFilter
 {
     private static final int MAX_RESOURCE_DEPTH = 3;
 
+    private final Function<RoleResource, Set<RoleResource>> getRolesFunction;
     private final AuditWhitelistCache whitelistCache;
     private final WhitelistDataAccess whitelistDataAccess;
 
     public RoleAuditFilter()
     {
-        this(AuditWhitelistCache.getInstance(), WhitelistDataAccess.getInstance());
+        this(Roles::getRoles, AuditWhitelistCache.getInstance(), WhitelistDataAccess.getInstance());
     }
 
     @VisibleForTesting
-    RoleAuditFilter(AuditWhitelistCache whitelistCache, WhitelistDataAccess whitelistDataAccess)
+    RoleAuditFilter(Function<RoleResource, Set<RoleResource>> getRolesFunction, AuditWhitelistCache whitelistCache, WhitelistDataAccess whitelistDataAccess)
     {
+        this.getRolesFunction = getRolesFunction;
         this.whitelistCache = whitelistCache;
         this.whitelistDataAccess = whitelistDataAccess;
     }
@@ -78,10 +80,9 @@ public class RoleAuditFilter implements AuditFilter
     @Override
     public boolean isFiltered(AuditEntry logEntry)
     {
-        RoleResource primaryRole = RoleResource.role(logEntry.getUser());
         try
         {
-            return isFiltered(Roles.getRoles(primaryRole), logEntry.getPermissions(), logEntry.getResource());
+            return isFilteredUnchecked(logEntry);
         }
         catch (UncheckedExecutionException e)
         {
@@ -89,11 +90,11 @@ public class RoleAuditFilter implements AuditFilter
         }
     }
 
-    @VisibleForTesting
-    boolean isFiltered(Set<RoleResource> roles, Set<Permission> operations, IResource resource)
+    private boolean isFilteredUnchecked(AuditEntry logEntry)
     {
-        List<IResource> operationResourceHierarchy = getResourceHierarchy(resource);
-        for (Permission operation : operations)
+        Set<RoleResource> roles = getRoles(logEntry.getUser());
+        List<IResource> operationResourceHierarchy = getResourceHierarchy(logEntry.getResource());
+        for (Permission operation : logEntry.getPermissions())
         {
             if(!isOperationWhitelistedOnResourceByRoles(operation, operationResourceHierarchy, roles))
             {
@@ -102,6 +103,12 @@ public class RoleAuditFilter implements AuditFilter
         }
 
         return true;
+    }
+
+    private Set<RoleResource> getRoles(String username)
+    {
+        RoleResource primaryRole = RoleResource.role(username);
+        return getRolesFunction.apply(primaryRole);
     }
 
     private List<IResource> getResourceHierarchy(IResource primaryResource)

--- a/ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/filter/role/RoleAuditFilter.java
+++ b/ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/filter/role/RoleAuditFilter.java
@@ -21,8 +21,10 @@ import java.util.Map;
 import java.util.Set;
 
 import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.UncheckedExecutionException;
 
 import com.ericsson.bss.cassandra.ecaudit.auth.AuditWhitelistCache;
+import com.ericsson.bss.cassandra.ecaudit.auth.Exceptions;
 import com.ericsson.bss.cassandra.ecaudit.auth.WhitelistDataAccess;
 import com.ericsson.bss.cassandra.ecaudit.entry.AuditEntry;
 import com.ericsson.bss.cassandra.ecaudit.filter.AuditFilter;
@@ -30,13 +32,14 @@ import org.apache.cassandra.auth.IResource;
 import org.apache.cassandra.auth.Permission;
 import org.apache.cassandra.auth.RoleResource;
 import org.apache.cassandra.auth.Roles;
+import org.apache.cassandra.exceptions.CassandraException;
 
 /**
  * A role based white-list filter that exempts users based on custom options on roles in Cassandra.
  */
 public class RoleAuditFilter implements AuditFilter
 {
-    public static final int MAX_RESOURCE_DEPTH = 3;
+    private static final int MAX_RESOURCE_DEPTH = 3;
 
     private final AuditWhitelistCache whitelistCache;
     private final WhitelistDataAccess whitelistDataAccess;
@@ -76,7 +79,14 @@ public class RoleAuditFilter implements AuditFilter
     public boolean isFiltered(AuditEntry logEntry)
     {
         RoleResource primaryRole = RoleResource.role(logEntry.getUser());
-        return isFiltered(Roles.getRoles(primaryRole), logEntry.getPermissions(), logEntry.getResource());
+        try
+        {
+            return isFiltered(Roles.getRoles(primaryRole), logEntry.getPermissions(), logEntry.getResource());
+        }
+        catch (UncheckedExecutionException e)
+        {
+            throw Exceptions.tryGetCassandraExceptionCause(e);
+        }
     }
 
     @VisibleForTesting

--- a/ecaudit/src/test/java/com/ericsson/bss/cassandra/ecaudit/TestAuditAdapter.java
+++ b/ecaudit/src/test/java/com/ericsson/bss/cassandra/ecaudit/TestAuditAdapter.java
@@ -53,8 +53,11 @@ import org.apache.cassandra.cql3.ColumnSpecification;
 import org.apache.cassandra.cql3.QueryOptions;
 import org.apache.cassandra.cql3.statements.BatchStatement;
 import org.apache.cassandra.cql3.statements.ModificationStatement;
+import org.apache.cassandra.db.ConsistencyLevel;
 import org.apache.cassandra.db.marshal.UTF8Type;
 import org.apache.cassandra.dht.IPartitioner;
+import org.apache.cassandra.exceptions.AuthenticationException;
+import org.apache.cassandra.exceptions.ReadTimeoutException;
 import org.apache.cassandra.service.ClientState;
 import org.apache.cassandra.utils.MD5Digest;
 import org.mockito.ArgumentCaptor;
@@ -63,8 +66,10 @@ import org.mockito.Mockito;
 import org.mockito.junit.MockitoJUnitRunner;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.doThrow;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.spy;
 import static org.mockito.Mockito.times;
@@ -491,6 +496,24 @@ public class TestAuditAdapter
         assertThat(captured.getPermissions()).isEqualTo(Sets.immutableEnumSet(Permission.EXECUTE));
         assertThat(captured.getResource()).isEqualTo(ConnectionResource.root());
         assertThat(captured.getTimestamp()).isEqualTo(TIMESTAMP);
+    }
+
+    @Test
+    public void testProcessAuthException()
+    {
+        InetAddress expectedAddress = mock(InetAddress.class);
+        String expectedUser = "user";
+        Status expectedStatus = Status.ATTEMPT;
+
+        when(mockAuditEntryBuilderFactory.createAuthenticationEntryBuilder())
+        .thenReturn(AuditEntry.newBuilder()
+                              .permissions(ImmutableSet.of(Permission.EXECUTE))
+                              .resource(ConnectionResource.root()));
+
+        doThrow(new ReadTimeoutException(ConsistencyLevel.QUORUM, 3, 4, true)).when(mockAuditor).audit(any(AuditEntry.class));
+
+        assertThatExceptionOfType(AuthenticationException.class)
+        .isThrownBy(() -> auditAdapter.auditAuth(expectedUser, expectedAddress, expectedStatus, TIMESTAMP));
     }
 
     private ImmutableList<ColumnSpecification> createTextColumns(String... columns)

--- a/ecaudit/src/test/java/com/ericsson/bss/cassandra/ecaudit/auth/TestExceptions.java
+++ b/ecaudit/src/test/java/com/ericsson/bss/cassandra/ecaudit/auth/TestExceptions.java
@@ -1,0 +1,88 @@
+/*
+ * Copyright 2019 Telefonaktiebolaget LM Ericsson
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.ericsson.bss.cassandra.ecaudit.auth;
+
+import org.junit.Test;
+
+import org.apache.cassandra.db.ConsistencyLevel;
+import org.apache.cassandra.exceptions.ReadTimeoutException;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestExceptions
+{
+    @Test
+    public void testSimpleRuntimeExceptionIsSame()
+    {
+        RuntimeException expectedException = new RuntimeException();
+
+        RuntimeException actualCause = Exceptions.tryGetCassandraExceptionCause(expectedException);
+
+        assertThat(actualCause).isSameAs(expectedException);
+        assertThat(actualCause.getSuppressed()).isEmpty();
+    }
+
+    @Test
+    public void testSimpleCassandraExceptionIsSame()
+    {
+        RuntimeException expectedException = new ReadTimeoutException(ConsistencyLevel.QUORUM, 1, 1, false);
+
+        RuntimeException actualCause = Exceptions.tryGetCassandraExceptionCause(expectedException);
+
+        assertThat(actualCause).isSameAs(expectedException);
+        assertThat(actualCause.getSuppressed()).isEmpty();
+    }
+
+    @Test
+    public void testWrappedRuntimeExceptionIsSame()
+    {
+        RuntimeException exception = new RuntimeException();
+        RuntimeException expectedException = new RuntimeException(exception);
+
+        RuntimeException actualCause = Exceptions.tryGetCassandraExceptionCause(expectedException);
+
+        assertThat(actualCause).isSameAs(expectedException);
+        assertThat(actualCause.getSuppressed()).isEmpty();
+        assertThat(exception.getSuppressed()).isEmpty();
+    }
+
+    @Test
+    public void testWrappedCassandraExceptionIsUnwrapped()
+    {
+        RuntimeException expectedException = new ReadTimeoutException(ConsistencyLevel.QUORUM, 1, 1, false);
+        RuntimeException exception = new RuntimeException(expectedException);
+
+        RuntimeException actualCause = Exceptions.tryGetCassandraExceptionCause(exception);
+
+        assertThat(actualCause).isSameAs(expectedException);
+        assertThat(actualCause.getSuppressed()).contains(exception);
+        assertThat(exception.getSuppressed()).isEmpty();
+    }
+
+    @Test
+    public void testMultiWrappedCassandraExceptionIsUnwrapped()
+    {
+        RuntimeException expectedException = new ReadTimeoutException(ConsistencyLevel.QUORUM, 1, 1, false);
+        RuntimeException exception = new RuntimeException(expectedException);
+        exception = new RuntimeException(exception);
+
+        RuntimeException actualCause = Exceptions.tryGetCassandraExceptionCause(exception);
+
+        assertThat(actualCause).isSameAs(expectedException);
+        assertThat(actualCause.getSuppressed()).contains(exception);
+        assertThat(exception.getSuppressed()).isEmpty();
+    }
+}

--- a/ecaudit/src/test/java/com/ericsson/bss/cassandra/ecaudit/auth/TestSchemaHelper.java
+++ b/ecaudit/src/test/java/com/ericsson/bss/cassandra/ecaudit/auth/TestSchemaHelper.java
@@ -16,6 +16,7 @@
 package com.ericsson.bss.cassandra.ecaudit.auth;
 
 import java.net.InetAddress;
+import java.net.UnknownHostException;
 import java.util.UUID;
 
 import com.google.common.collect.ImmutableSet;
@@ -41,15 +42,15 @@ import static org.mockito.Mockito.when;
 @RunWith(MockitoJUnitRunner.StrictStubs.class)
 public class TestSchemaHelper
 {
+    private static final long RETRY_INTERVAL_MILLIS = 100;
+
     @Mock
     private Gossiper mockGossiper;
 
     private SchemaHelper schemaHelper;
 
-    private EndpointState localEndpointState;
-    private EndpointState anotherEndpointState;
-
-    private InetAddress remoteEndpoint;
+    private EndpointState currentEndpointState;
+    private EndpointState previousEndpointState;
 
     @BeforeClass
     public static void beforeClass()
@@ -60,17 +61,14 @@ public class TestSchemaHelper
     @Before
     public void before() throws Exception
     {
+        currentEndpointState = createEndpointStateWithSchema();
+        previousEndpointState = createEndpointStateWithSchema();
+
         InetAddress localEndpoint = InetAddress.getLoopbackAddress();
-        localEndpointState = createEndpointStateWithSchema();
-        when(mockGossiper.getEndpointStateForEndpoint(eq(localEndpoint))).thenReturn(localEndpointState);
-
-        remoteEndpoint = InetAddress.getByName("127.0.0.2");
-
-        anotherEndpointState = createEndpointStateWithSchema();
-
+        InetAddress remoteEndpoint = InetAddress.getByName("127.0.0.2");
         when(mockGossiper.getLiveMembers()).thenReturn(ImmutableSet.of(localEndpoint, remoteEndpoint));
 
-        schemaHelper = new SchemaHelper(localEndpoint, mockGossiper);
+        schemaHelper = new SchemaHelper(localEndpoint, mockGossiper, RETRY_INTERVAL_MILLIS);
     }
 
     @AfterClass
@@ -79,40 +77,66 @@ public class TestSchemaHelper
         Config.setClientMode(false);
     }
 
-    @Test(timeout = 999)
-    public void testSyncedFromStart()
+    @Test(timeout = (RETRY_INTERVAL_MILLIS) - 1)
+    public void testSyncedFromStart() throws UnknownHostException
     {
-        when(mockGossiper.getEndpointStateForEndpoint(eq(remoteEndpoint)))
-        .thenReturn(localEndpointState);
+        givenLocalStateOrder(currentEndpointState);
+        givenRemoteStateOrder(currentEndpointState);
 
-        schemaHelper.waitForSchemaAlignment(10);
+        assertThat(schemaHelper.areSchemasAligned(10 * RETRY_INTERVAL_MILLIS)).isTrue();
     }
 
-    @Test(timeout = 1999)
-    public void testSyncedAfterOneSecond()
+    @Test(timeout = (3 * RETRY_INTERVAL_MILLIS) - 1)
+    public void testRemoveSyncedAfterTwoTries() throws UnknownHostException
     {
-        when(mockGossiper.getEndpointStateForEndpoint(eq(remoteEndpoint)))
-        .thenReturn(anotherEndpointState)
-        .thenReturn(localEndpointState);
+        givenLocalStateOrder(currentEndpointState);
+        givenRemoteStateOrder(null, previousEndpointState, currentEndpointState);
 
         long startTime = System.currentTimeMillis();
 
-        schemaHelper.waitForSchemaAlignment(10);
+        assertThat(schemaHelper.areSchemasAligned(10 * RETRY_INTERVAL_MILLIS)).isTrue();
 
-        assertWaitTimeInMillis(startTime, 1000L);
+        assertWaitTimeInMillis(startTime, 2 * RETRY_INTERVAL_MILLIS);
     }
 
-    @Test(timeout = 1999)
-    public void testSyncTimeout()
+    @Test(timeout = (2 * RETRY_INTERVAL_MILLIS) - 1)
+    public void testSyncTimeout() throws UnknownHostException
     {
-        when(mockGossiper.getEndpointStateForEndpoint(eq(remoteEndpoint)))
-        .thenReturn(anotherEndpointState);
+        givenLocalStateOrder(currentEndpointState);
+        givenRemoteStateOrder(previousEndpointState);
 
         long startTime = System.currentTimeMillis();
 
-        schemaHelper.waitForSchemaAlignment(1);
+        assertThat(schemaHelper.areSchemasAligned(RETRY_INTERVAL_MILLIS)).isFalse();
 
-        assertWaitTimeInMillis(startTime, 1000L);
+        assertWaitTimeInMillis(startTime, RETRY_INTERVAL_MILLIS);
+    }
+
+    @Test(timeout = (2 * RETRY_INTERVAL_MILLIS) - 1)
+    public void testNoLocalSchemaInFirstTry() throws UnknownHostException
+    {
+        givenLocalStateOrder(null, currentEndpointState);
+        givenRemoteStateOrder(currentEndpointState);
+
+        long startTime = System.currentTimeMillis();
+
+        assertThat(schemaHelper.areSchemasAligned(10 * RETRY_INTERVAL_MILLIS)).isTrue();
+
+        assertWaitTimeInMillis(startTime, RETRY_INTERVAL_MILLIS);
+    }
+
+    private void givenLocalStateOrder(EndpointState endpointState, EndpointState... endpointStates)
+    {
+        InetAddress localEndpoint = InetAddress.getLoopbackAddress();
+        when(mockGossiper.getEndpointStateForEndpoint(eq(localEndpoint)))
+        .thenReturn(endpointState, endpointStates);
+    }
+
+    private void givenRemoteStateOrder(EndpointState endpointState, EndpointState... endpointStates) throws UnknownHostException
+    {
+        InetAddress remoteEndpoint = InetAddress.getByName("127.0.0.2");
+        when(mockGossiper.getEndpointStateForEndpoint(eq(remoteEndpoint)))
+        .thenReturn(endpointState, endpointStates);
     }
 
     private EndpointState createEndpointStateWithSchema()
@@ -125,9 +149,9 @@ public class TestSchemaHelper
         return endpointState;
     }
 
-    private void assertWaitTimeInMillis(long startTime, long delay)
+    private void assertWaitTimeInMillis(long startTime, long millis)
     {
         long now = System.currentTimeMillis();
-        assertThat(now - startTime).isGreaterThan(delay);
+        assertThat(now - startTime).isGreaterThanOrEqualTo(millis);
     }
 }

--- a/ecaudit/src/test/java/com/ericsson/bss/cassandra/ecaudit/auth/TestSchemaHelper.java
+++ b/ecaudit/src/test/java/com/ericsson/bss/cassandra/ecaudit/auth/TestSchemaHelper.java
@@ -1,0 +1,133 @@
+/*
+ * Copyright 2019 Telefonaktiebolaget LM Ericsson
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.ericsson.bss.cassandra.ecaudit.auth;
+
+import java.net.InetAddress;
+import java.util.UUID;
+
+import com.google.common.collect.ImmutableSet;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import org.apache.cassandra.config.Config;
+import org.apache.cassandra.gms.ApplicationState;
+import org.apache.cassandra.gms.EndpointState;
+import org.apache.cassandra.gms.Gossiper;
+import org.apache.cassandra.gms.VersionedValue;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.StrictStubs.class)
+public class TestSchemaHelper
+{
+    @Mock
+    private Gossiper mockGossiper;
+
+    private SchemaHelper schemaHelper;
+
+    private EndpointState localEndpointState;
+    private EndpointState anotherEndpointState;
+
+    private InetAddress remoteEndpoint;
+
+    @BeforeClass
+    public static void beforeClass()
+    {
+        Config.setClientMode(true);
+    }
+
+    @Before
+    public void before() throws Exception
+    {
+        InetAddress localEndpoint = InetAddress.getLoopbackAddress();
+        localEndpointState = createEndpointStateWithSchema();
+        when(mockGossiper.getEndpointStateForEndpoint(eq(localEndpoint))).thenReturn(localEndpointState);
+
+        remoteEndpoint = InetAddress.getByName("127.0.0.2");
+
+        anotherEndpointState = createEndpointStateWithSchema();
+
+        when(mockGossiper.getLiveMembers()).thenReturn(ImmutableSet.of(localEndpoint, remoteEndpoint));
+
+        schemaHelper = new SchemaHelper(localEndpoint, mockGossiper);
+    }
+
+    @AfterClass
+    public static void afterClass()
+    {
+        Config.setClientMode(false);
+    }
+
+    @Test(timeout = 999)
+    public void testSyncedFromStart()
+    {
+        when(mockGossiper.getEndpointStateForEndpoint(eq(remoteEndpoint)))
+        .thenReturn(localEndpointState);
+
+        schemaHelper.waitForSchemaAlignment(10);
+    }
+
+    @Test(timeout = 1999)
+    public void testSyncedAfterOneSecond()
+    {
+        when(mockGossiper.getEndpointStateForEndpoint(eq(remoteEndpoint)))
+        .thenReturn(anotherEndpointState)
+        .thenReturn(localEndpointState);
+
+        long startTime = System.currentTimeMillis();
+
+        schemaHelper.waitForSchemaAlignment(10);
+
+        assertWaitTimeInMillis(startTime, 1000L);
+    }
+
+    @Test(timeout = 1999)
+    public void testSyncTimeout()
+    {
+        when(mockGossiper.getEndpointStateForEndpoint(eq(remoteEndpoint)))
+        .thenReturn(anotherEndpointState);
+
+        long startTime = System.currentTimeMillis();
+
+        schemaHelper.waitForSchemaAlignment(1);
+
+        assertWaitTimeInMillis(startTime, 1000L);
+    }
+
+    private EndpointState createEndpointStateWithSchema()
+    {
+        VersionedValue versionedValue = new VersionedValue.VersionedValueFactory(null).schema(UUID.randomUUID());
+
+        EndpointState endpointState = mock(EndpointState.class);
+        when(endpointState.getApplicationState(eq(ApplicationState.SCHEMA))).thenReturn(versionedValue);
+
+        return endpointState;
+    }
+
+    private void assertWaitTimeInMillis(long startTime, long delay)
+    {
+        long now = System.currentTimeMillis();
+        assertThat(now - startTime).isGreaterThan(delay);
+    }
+}

--- a/ecaudit/src/test/java/com/ericsson/bss/cassandra/ecaudit/filter/role/TestRoleAuditFilter.java
+++ b/ecaudit/src/test/java/com/ericsson/bss/cassandra/ecaudit/filter/role/TestRoleAuditFilter.java
@@ -19,10 +19,12 @@ import java.util.Arrays;
 import java.util.Collections;
 import java.util.Map;
 import java.util.Set;
+import java.util.function.Function;
 import java.util.stream.Collectors;
 
 import com.google.common.collect.Maps;
 import com.google.common.collect.Sets;
+import com.google.common.util.concurrent.UncheckedExecutionException;
 import org.junit.Before;
 import org.junit.Test;
 import org.junit.runner.RunWith;
@@ -30,14 +32,19 @@ import org.junit.runner.RunWith;
 import com.ericsson.bss.cassandra.ecaudit.auth.AuditWhitelistCache;
 import com.ericsson.bss.cassandra.ecaudit.auth.ConnectionResource;
 import com.ericsson.bss.cassandra.ecaudit.auth.WhitelistDataAccess;
+import com.ericsson.bss.cassandra.ecaudit.entry.AuditEntry;
 import org.apache.cassandra.auth.DataResource;
 import org.apache.cassandra.auth.IResource;
 import org.apache.cassandra.auth.Permission;
 import org.apache.cassandra.auth.RoleResource;
+import org.apache.cassandra.db.ConsistencyLevel;
+import org.apache.cassandra.exceptions.CassandraException;
+import org.apache.cassandra.exceptions.ReadTimeoutException;
 import org.mockito.Mock;
 import org.mockito.junit.MockitoJUnitRunner;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.Mockito.times;
 import static org.mockito.Mockito.verify;
@@ -46,6 +53,9 @@ import static org.mockito.Mockito.when;
 @RunWith(MockitoJUnitRunner.StrictStubs.class)
 public class TestRoleAuditFilter
 {
+    @Mock
+    private Function<RoleResource, Set<RoleResource>> getRolesFunctionMock;
+
     @Mock
     private AuditWhitelistCache auditWhitelistCacheMock;
     private Map<RoleResource, Map<IResource, Set<Permission>>> whitelistMap;
@@ -58,7 +68,7 @@ public class TestRoleAuditFilter
     @Before
     public void before()
     {
-        filter = new RoleAuditFilter(auditWhitelistCacheMock, whitelistDataAccessMock);
+        filter = new RoleAuditFilter(getRolesFunctionMock, auditWhitelistCacheMock, whitelistDataAccessMock);
 
         whitelistMap = Maps.newHashMap();
         when(auditWhitelistCacheMock.getWhitelist(any(RoleResource.class)))
@@ -79,32 +89,40 @@ public class TestRoleAuditFilter
     public void primaryRoleWithWhitelistedDataRootDoSelect()
     {
         givenRoleIsWhitelisted("primary", Permission.SELECT, DataResource.root());
-        Set<RoleResource> effectiveRoles = givenRolesOfRequest("primary", "inherited");
-        assertThat(filter.isFiltered(effectiveRoles, Collections.singleton(Permission.SELECT), DataResource.fromName("data/ks/tbl"))).isEqualTo(true);
+        givenRolesOfRequest("primary", "inherited");
+        AuditEntry auditEntry = givenAuditEntry(Collections.singleton(Permission.SELECT), DataResource.fromName("data/ks/tbl"));
+
+        assertThat(filter.isFiltered(auditEntry)).isTrue();
     }
 
     @Test
     public void primaryRoleWithWhitelistedKsDoSelect()
     {
         givenRoleIsWhitelisted("primary", Permission.SELECT, DataResource.fromName("data/ks"));
-        Set<RoleResource> effectiveRoles = givenRolesOfRequest("primary", "inherited");
-        assertThat(filter.isFiltered(effectiveRoles, Collections.singleton(Permission.SELECT), DataResource.fromName("data/ks/tbl"))).isEqualTo(true);
+        givenRolesOfRequest("primary", "inherited");
+        AuditEntry auditEntry = givenAuditEntry(Collections.singleton(Permission.SELECT), DataResource.fromName("data/ks/tbl"));
+
+        assertThat(filter.isFiltered(auditEntry)).isTrue();
     }
 
     @Test
     public void primaryRoleWithWhitelistedTableDoSelect()
     {
         givenRoleIsWhitelisted("primary", Permission.SELECT, DataResource.fromName("data/ks/tbl"));
-        Set<RoleResource> effectiveRoles = givenRolesOfRequest("primary", "inherited");
-        assertThat(filter.isFiltered(effectiveRoles, Collections.singleton(Permission.SELECT), DataResource.fromName("data/ks/tbl"))).isEqualTo(true);
+        givenRolesOfRequest("primary", "inherited");
+        AuditEntry auditEntry = givenAuditEntry(Collections.singleton(Permission.SELECT), DataResource.fromName("data/ks/tbl"));
+
+        assertThat(filter.isFiltered(auditEntry)).isTrue();
     }
 
     @Test
     public void primaryRoleWithWhitelistedTableDoModify()
     {
         givenRoleIsWhitelisted("primary", Permission.MODIFY, DataResource.fromName("data/ks/tbl"));
-        Set<RoleResource> effectiveRoles = givenRolesOfRequest("primary", "inherited");
-        assertThat(filter.isFiltered(effectiveRoles, Collections.singleton(Permission.MODIFY), DataResource.fromName("data/ks/tbl"))).isEqualTo(true);
+        givenRolesOfRequest("primary", "inherited");
+        AuditEntry auditEntry = givenAuditEntry(Collections.singleton(Permission.MODIFY), DataResource.fromName("data/ks/tbl"));
+
+        assertThat(filter.isFiltered(auditEntry)).isTrue();
     }
 
     @Test
@@ -112,8 +130,10 @@ public class TestRoleAuditFilter
     {
         givenRoleIsWhitelisted("primary", Permission.SELECT, DataResource.fromName("data/ks/tbl"));
         givenRoleIsWhitelisted("primary", Permission.MODIFY, DataResource.fromName("data/ks/tbl"));
-        Set<RoleResource> effectiveRoles = givenRolesOfRequest("primary", "inherited");
-        assertThat(filter.isFiltered(effectiveRoles, Sets.newHashSet(Permission.SELECT, Permission.MODIFY), DataResource.fromName("data/ks/tbl"))).isEqualTo(true);
+        givenRolesOfRequest("primary", "inherited");
+        AuditEntry auditEntry = givenAuditEntry(Sets.newHashSet(Permission.SELECT, Permission.MODIFY), DataResource.fromName("data/ks/tbl"));
+
+        assertThat(filter.isFiltered(auditEntry)).isTrue();
     }
 
     @Test
@@ -121,53 +141,78 @@ public class TestRoleAuditFilter
     {
         givenRoleIsWhitelisted("primary", Permission.SELECT, DataResource.fromName("data/ks/tbl"));
         givenRoleIsWhitelisted("inherited", Permission.MODIFY, DataResource.fromName("data/ks/tbl"));
-        Set<RoleResource> effectiveRoles = givenRolesOfRequest("primary", "inherited");
-        assertThat(filter.isFiltered(effectiveRoles, Sets.newHashSet(Permission.SELECT, Permission.MODIFY), DataResource.fromName("data/ks/tbl"))).isEqualTo(true);
+        givenRolesOfRequest("primary", "inherited");
+        AuditEntry auditEntry = givenAuditEntry(Sets.newHashSet(Permission.SELECT, Permission.MODIFY), DataResource.fromName("data/ks/tbl"));
+
+        assertThat(filter.isFiltered(auditEntry)).isTrue();
     }
 
     @Test
     public void primaryRoleWithOtherWhitelistedTableDoSelect()
     {
         givenRoleIsWhitelisted("primary", Permission.SELECT, DataResource.fromName("data/ks/tbl"));
-        Set<RoleResource> effectiveRoles = givenRolesOfRequest("primary", "inherited");
-        assertThat(filter.isFiltered(effectiveRoles, Collections.singleton(Permission.SELECT), DataResource.fromName("data/ks/other_tbl"))).isEqualTo(false);
+        givenRolesOfRequest("primary", "inherited");
+        AuditEntry auditEntry = givenAuditEntry(Collections.singleton(Permission.SELECT), DataResource.fromName("data/ks/other_tbl"));
+
+        assertThat(filter.isFiltered(auditEntry)).isFalse();
     }
 
     @Test
     public void inheritedRoleWithWhitelistedDataRootDoSelect()
     {
         givenRoleIsWhitelisted("inherited", Permission.SELECT, DataResource.root());
-        Set<RoleResource> effectiveRoles = givenRolesOfRequest("primary", "inherited");
-        assertThat(filter.isFiltered(effectiveRoles, Collections.singleton(Permission.SELECT), DataResource.fromName("data/ks/tbl"))).isEqualTo(true);
+        givenRolesOfRequest("primary", "inherited");
+        AuditEntry auditEntry = givenAuditEntry(Collections.singleton(Permission.SELECT), DataResource.fromName("data/ks/tbl"));
+
+        assertThat(filter.isFiltered(auditEntry)).isTrue();
     }
 
     @Test
     public void inheritedRoleWithWhitelistedConnectionDoConnect()
     {
         givenRoleIsWhitelisted("inherited", Permission.EXECUTE, ConnectionResource.root());
-        Set<RoleResource> effectiveRoles = givenRolesOfRequest("primary", "inherited");
-        assertThat(filter.isFiltered(effectiveRoles, Collections.singleton(Permission.EXECUTE), ConnectionResource.root())).isEqualTo(true);
+        givenRolesOfRequest("primary", "inherited");
+        AuditEntry auditEntry = givenAuditEntry(Collections.singleton(Permission.EXECUTE), ConnectionResource.root());
+
+        assertThat(filter.isFiltered(auditEntry)).isTrue();
     }
 
     @Test
     public void roleDoConnect()
     {
-        Set<RoleResource> effectiveRoles = givenRolesOfRequest("primary", "inherited");
-        assertThat(filter.isFiltered(effectiveRoles, Collections.singleton(Permission.EXECUTE), ConnectionResource.root())).isEqualTo(false);
+        givenRolesOfRequest("primary", "inherited");
+        AuditEntry auditEntry = givenAuditEntry(Collections.singleton(Permission.EXECUTE), ConnectionResource.root());
+
+        assertThat(filter.isFiltered(auditEntry)).isFalse();
     }
 
     @Test
     public void roleDoModify()
     {
-        Set<RoleResource> effectiveRoles = givenRolesOfRequest("primary", "inherited");
-        assertThat(filter.isFiltered(effectiveRoles, Collections.singleton(Permission.MODIFY), DataResource.fromName("data/ks/tbl"))).isEqualTo(false);
+        givenRolesOfRequest("primary", "inherited");
+        AuditEntry auditEntry = givenAuditEntry(Collections.singleton(Permission.MODIFY), DataResource.fromName("data/ks/tbl"));
+
+        assertThat(filter.isFiltered(auditEntry)).isFalse();
     }
 
-    private Set<RoleResource> givenRolesOfRequest(String... roleNames)
+    @Test
+    public void uncheckedExceptionIsUnwrapped()
     {
-        return Arrays.stream(roleNames)
-                     .map(RoleResource::role)
-                     .collect(Collectors.toSet());
+        when(getRolesFunctionMock.apply(any(RoleResource.class)))
+        .thenThrow(new UncheckedExecutionException(new RuntimeException(new ReadTimeoutException(ConsistencyLevel.QUORUM, 1, 1, false))));
+        AuditEntry auditEntry = givenAuditEntry(Collections.singleton(Permission.SELECT), DataResource.fromName("data/ks/tbl"));
+
+        assertThatExceptionOfType(CassandraException.class)
+        .isThrownBy(() -> filter.isFiltered(auditEntry));
+    }
+
+    private void givenRolesOfRequest(String... roleNames)
+    {
+        Set<RoleResource> roles = Arrays.stream(roleNames)
+                                        .map(RoleResource::role)
+                                        .collect(Collectors.toSet());
+
+        when(getRolesFunctionMock.apply(any(RoleResource.class))).thenReturn(roles);
     }
 
     private void givenRoleIsWhitelisted(String roleName, Permission operation, IResource resource)
@@ -187,5 +232,13 @@ public class TestRoleAuditFilter
         Set<Permission> newPermissionSet = permissions != null ? permissions : Sets.newHashSet();
         newPermissionSet.add(permission);
         return newPermissionSet;
+    }
+
+    private AuditEntry givenAuditEntry(Set<Permission> permissions, IResource resource)
+    {
+        return AuditEntry.newBuilder()
+                         .permissions(permissions)
+                         .resource(resource)
+                         .build();
     }
 }

